### ISSUE TITLE
Unable to download spreadsheet from ingest endpoint 

### ISSUE THREAD

**Issue by Wkt8** [ebi-ait/dcp-ingest-central#491]
**Describe the bug**
> Unable to download the up-to-date version of a spreadsheet from spreadsheet endpoint in ingest for a project

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to https://contribute.data.humancellatlas.org/projects/detail?uuid=6c040a93-8cf8-4fd5-98de-2297eb07e9f6
2. Move to the 'Data upload' tab
3. Click on the existing submission 
4. Click on the spreadsheet endpoint to download the spreadsheet (next to the 'back to project' page) 
5. The screen is stuck and the spreadsheet does not download 

**Expected behaviour**
When clicking on the spreadsheet endpoint, it should download an up-to-date version of the spreadsheet for that submission 

**Screenshots / Logs**
> If applicable, add screenshots or log files to help explain your problem.

**Links**
Other Examples
https://contribute.data.humancellatlas.org/projects/detail?uuid=6c040a93-8cf8-4fd5-98de-2297eb07e9f6
https://contribute.data.humancellatlas.org/projects/detail?uuid=71eb5f6d-cee0-4297-b503-b1125909b8c7
https://contribute.data.humancellatlas.org/projects/detail?uuid=c4077b3c-5c98-4d26-a614-246d12c2e5d7

**Environment**
- [x] prod
- [ ] staging
- [ ] dev

**Browser**
- [x] Chrome
- [ ] Safari
- [ ] IE
- [ ] Other (please specify)

**Comment by Wkt8** [comment]
Enrique mentioned it wasn't working for him last thursday either. It is also not working for some of the UCSC wranglers. Blocking https://github.com/ebi-ait/hca-ebi-wrangler-central/issues/513

**Comment by aaclan-ebi** [comment]
@Wkt8 please note that the linking columns (e.g. cell suspension id in the sequence file tab) will not be included in the downloaded spreadsheet. I believe you would need that to investigate the issue? ebi-ait/hca-ebi-wrangler-central#513

It might be better to get the original spreadsheet which should be available using this endpoint:
`https://ingest.archive.data.humancellatlas.org/submissions/<submission-uuid>/spreadsheet/original`

e.g.

https://ingest.archive.data.humancellatlas.org/submissions/6b12e130-d3d8-41b1-ae41-d0131e890fb8/spreadsheet/original

Please note that if linking is changed via UI, it wouldn't be reflected in the original spreadsheet :(

**Comment by aaclan-ebi** [comment]
<img width="1436" alt="Screenshot 2021-10-18 at 11 27 53" src="https://user-images.githubusercontent.com/31281398/137714244-36cce9b0-f680-4eeb-8fc3-b0b264842250.png">

The UI already displays an error while the broker is still processing the request. It's not clear what the error is because the response HTTP status is 0. I am guessing the UI timed out in processing the request because the process is still ongoing in the broker service. 

We might need to modify the implementation to make the request asynchronous and let the user download when spreadsheet generation is done.

**Comment by Wkt8** [comment]
Okay, thank you for the information! It does not display that error page for me when I attempt to download - but the original spreadsheet endpoint is very useful, thanks. 

The error seems to occur inconsistently - as I was just able to download spreadsheets from the ingest-endpoints of all of the different submissions. Is it still worth keeping the ticket open to fix the issue, even though it is inconsistent?

**Comment by aaclan-ebi** [comment]
Yup, this is still a big issue in the download spreadsheet feature and is definitely needed to be fixed.

It's just that this shouldn't block the ebi-ait/hca-ebi-wrangler-central#513 because you wouldn't be able to see the linking info from the downloaded spreadsheet which you would need to investigate that ticket.

The original spreadsheet could help and you could also navigate the API's as well to see how entities are linked (probably not the easiest way for a human user).

1 - if you go to the submission JSON url : https://api.ingest.archive.data.humancellatlas.org/submissionEnvelopes/601c0432ac1792031eee3f43

2 - and then if you check the _links.files.href from submission json
https://api.ingest.archive.data.humancellatlas.org/submissionEnvelopes/601c0432ac1792031eee3f43/files

3 - and then if you check _embedded.files.0._links.derivedByProcesses.href 
https://api.ingest.archive.data.humancellatlas.org/files/601c043bac1792031eee3f62/derivedByProcesses

that should give you the assay process resource

4 - and then you can check the _embedded.processes.0._links.inputBiomaterials.href
https://api.ingest.archive.data.humancellatlas.org/processes/601c043cac1792031eee3f8f/inputBiomaterials

to see the input biomaterial to that assay process.

We could also file a ticket to support the ability to easily view how entities are linked. It could be by including the linking columns in the downloaded spreadsheet or some other way in the UI (viewing the graph per assay). That needs design.

**Comment by Wkt8** [comment]
ebi-ait/hca-ebi-wrangler-central#513 is no longer an issue as it is mostly resolved now and doesn't really requiring investigation of the submissions. Thank you for the info!

**Comment by amnonkhen** [comment]
Could we increase the timeout the browser uses for the request?

**Comment by ESapenaVentura** [comment]
3 options considered - Need to decide on which one @amnonkhen @aaclan-ebi

**Comment by ESapenaVentura** [comment]
TBC for next sprint

**Comment by prabh-t** [comment]
I tested this yesterday and again this morning but could not reproduce - I was able to download the spreadsheet for all the listed submissions above. @ESapenaVentura could you please comment on the 3 options considered?

**Comment by gabsie** [comment]
@prabh-t Prabhat, are you working on this? Just checking whether it's you or Alegria..  :)

**Comment by prabh-t** [comment]
Hi Gabs! yeps, have been trying some tricks and discussing solutions with the crew.

**Comment by ipediez** [comment]
Generating spreadsheets is taking too long, we need to take a look at this issue. @prabh-t to organise a discussion

**Comment by prabh-t** [comment]
PR for fix as discussed w/ @amnonkhen https://github.com/ebi-ait/ingest-ui/pull/138

**Comment by prabh-t** [comment]
- Please wait...

![image.png](https://images.zenhubusercontent.com/5cff7550f579cd2bc2eb74b8/45902455-dd97-424f-bdce-de805845fba6)

- Download spreadsheet request times out

![image.png](https://images.zenhubusercontent.com/5cff7550f579cd2bc2eb74b8/30226264-4ca9-4692-9620-9fd534a9499c)

- Download button reactivates after a backoff period (20 mins) to prevent successive requests/overloading the cluster when a request times out.

![image.png](https://images.zenhubusercontent.com/5cff7550f579cd2bc2eb74b8/6866d44b-8db3-4b6b-bf56-c3f4548dd042)

**Comment by aaclan-ebi** [comment]
@prabh-t will add Angular unit tests on the changes.

**Comment by aaclan-ebi** [comment]
I strongly suggest that we go for the async way to solve this issue. Generating the spreadsheet can take a few mins especially for bigger datasets.

**Comment by jacobwindsor** [comment]
To replicate the issue:

1. Checkout `ingest-core` from before the revert
2. Run `ingest-core`, `ingest-UI`, and `ingest-broker`
3. Upload the SS2 integration test submission
4. You will see the error
5. To replicate the specific problem. You can:
  1. Locate the process that is not linked to the project (probably the last one)
  2. `curl -X PUT --data 'http://localhost:8080/projects/<project_id>' http://localhost:8080/processes/<process_id>/project -H "content-type: text/uri-list"

**Comment by idazucchi** [comment]
missing unit tests for the UI

**Comment by gabsie** [comment]
Let's discuss actually if there isn't another task here

**Comment by ofanobilbao** [comment]
@aaclan-ebi and @prabh-t worked on this yesterday. Prabhat will try to continue to work on this

**Comment by gabsie** [comment]
hey @prabh-t  - could you update on this, if there was any progress on the user notification bit?
Gabs

**Comment by prabh-t** [comment]
Hi Gabs, Alegria and I spent Wednesday afternoon adding tests to this change. I'll deploy to dev and staging later today.

**Comment by ke4** [comment]
@prabh-t continue working on this.

**Comment by prabh-t** [comment]
@ESapenaVentura to test today

**Comment by ESapenaVentura** [comment]
Hi @prabh-t I have tested with 2 submissions:

- Small one (SS2 integration test): Works as intended, I get the spreadsheet
- Intermediate one ([MouseArchitecture](https://staging.contribute.data.humancellatlas.org/submissions/detail?uuid=8d2a0ea9-3286-4c69-b6e7-16d861a5e23a)
   - Got a service unavailable error. Can't download the spreadsheet

- Large one ([Tabula muris](https://staging.contribute.data.humancellatlas.org/submissions/detail?uuid=14df1f92-155c-4da2-97fc-85601dee64da&project=e0009214-c0a0-4a7b-96e2-d6a83e966ce0))
   - I got a service unavailable error. Can't download the spreadsheet at all


I have tried several times the big and medium one, with no avail. Find attached a screenshot of the error:

![Screenshot 2021-12-20 at 12.00.12.png](https://images.zenhubusercontent.com/5ca5f897027e3e7234ac788c/13d7855d-10f7-407c-86ff-60e5f339e641)

**Comment by prabh-t** [comment]
@ESapenaVentura 
I was able to download the intermediate one. It took about 2 and a half minutes. 
I was not able to download the large one. this is not unexpected. This will probably take longer than the 10 mins we have currently set the new time out interval to be. However, it did time out only after 6 mins, not sure why.
The scope of the fix was to increase the time out interval to allow spreadsheets that were taking slightly longer to generate to completion (previous default timeout was 1 min).
The other thing the fix does is prevent the user from submitting another download request within 20 mins of initiating the first request. This is because if the spreadsheet is taking a long time (and resources) to generate in the backend, it will avoid spawning multiple duplicate processes in the cluster - which previously led to the pod being evicted. This is now slightly "controlled". However, the real problem of why spreadsheet generation takes sooo long is not tackled. This is another issue, that has to do with the inherent design of ingest api and/or spreadsheet generation code. I think there is another ticket for this.
The main goal of this one was to allow most not all spreadsheets to be able to be generated by increasing the time out interval and disabling multiple download requests in the frontend. The large spreadsheet will likely continue to present a problem as it is, until we review the whole problem/design.

**Comment by prabh-t** [comment]
It's not ok though that you are not able to download the intermediate one. How long does it take to show the error?

**Comment by ESapenaVentura** [comment]
@prabh-t I wasn't here when this ticket was discussed so I didn't know the scope of the ticket, sorry about that

So, the expected behaviour is that the screen is greyed out? (That's what is happening to me, I can see that below the overlay there is a greyed out button as well but I can't click anywhere on the screen anyways)

If that is what was expected... Are you expecting the button to be greyed out even if someone refreshes the page? or not?

Because I tried to refresh the page, click the download button, refresh again, click again... a couple of times in quick succession. Is that expected in the `disabling multiple download requests in the frontend` or is this bypassing that limitation that we try to address?

About your second comment, I did not time it but I can say that 
- First time I got it immediately
- Second time it took more than 5 minutes
- Third time it took less than second time

Would it help to know exactly how much time does it take? if it helps I can test it out

**Comment by prabh-t** [comment]
> If that is what was expected... Are you expecting the button to be greyed out even if someone refreshes the page?

Yes, this user should not be able to send another download request until the current one has finished, or 20mins, whichever is sooner. If you navigate away from the "please wait", and attempt to press download again, you should not be able to. This prevents the problem as I described above as the existing spreadsheet generation may still be running in the backend.

> Because I tried to refresh the page, click the download button, refresh again, click again... a couple of times in quick succession. Is that expected in the disabling multiple download requests in the frontend or is this bypassing that limitation that we try to address?

That should not be the case. If you are able to do that, then something is wrong. Looking into this.

**Comment by MightyAx** [comment]
We might already be as done as we can on this, unknown without more discussion between @gabsie & @amnonkhen

**Comment by gabsie** [comment]
This ticket seems done, but @prabh-t needs to double-check the deployment.
The result of the discussion is a new ticket #555. 
Needs a brief discussion @prabh-t @aaclan-ebi @gabsie

**Comment by ofanobilbao** [comment]
It was agreed that this can be closed after Prabhat did some changes.

**Comment by ofanobilbao** [comment]
@prabh-t to merge back to dev and deploy to prod the remaining

**Comment by ipediez** [comment]
Deployed on staging, dealing with last issues
